OCPBUGS-85045: Updated nw-using-cookies-keep-route-statefulness.adoc … (#111549)
openshift/openshift-docs: modules/nw-using-cookies-keep-route-statefulness.adoc

= Using cookies to keep route statefulness

[role="_abstract"]
To maintain stateful application traffic during pod restarts or scaling events, configure sticky sessions by using cookies. By using this method, you ensure that all incoming traffic reaches the same endpoint. This method helps prevent state loss even if the specific endpoint pod changes.

[IMPORTANT]
====

> The persistence cookie is enabled by default. For a response that contains the persistence cookie, the OpenShift router automatically appends the `Cache-Control: private` header to the response. This configuration prevents the cookie from being cached. If the persistence cookie was not in place, clients receiving the cached response would be routed to the same pod, resulting in performance and reliability issues.

Reviewer: @aaradhak good catch, you are correct. The paragraph should justify what would happen if the cookie is present (because the user hadn't disabled them) but `Cache-Control: private` header was not added.

= Using cookies to keep route statefulness

[role="_abstract"]
To maintain stateful application traffic during pod restarts or scaling events, configure sticky sessions by using cookies. By using this method, you ensure that all incoming traffic reaches the same endpoint. This method helps prevent state loss even if the specific endpoint pod changes.

[IMPORTANT]
====

> The persistence cookie is enabled by default. Responses that contain the persistence cookie must include the `Cache-Control: private` header to prevent the cookie from being cached. 

Reviewer: ```suggestion
The persistence cookie is enabled by default. Responses that contain the persistence cookie are automatically appended the `Cache-Control: private` header to prevent the cookie from being cached. 
```

Reviewer: Previous text looked like it is the user who is required to add the header, while we are the ones automatically and forcefully adding it.

Author: :heavy_check_mark:

> If the persistence cookie is cached, all clients receiving the cached response are routed to the same pod. This behavior can cause performance and reliability issues. 

Reviewer: ```suggestion
We need to append this header because, if the persistence cookie is cached, all clients receiving the cached response are routed to the same pod. This behavior can cause performance and reliability issues. 
```

Reviewer: The paragraph is a justification on why we forcefully add the header, not why we require the customer to do it (because we don't, it is us forcing it). I rewrote it at bit to make it clearer, but you might even want to fully re-phrase the paragraph in conditional (I'll let you decide here, as we would be entering the "text styling" area).

Author: :heavy_check_mark:

> Applications serving static content with custom `Cache-Control` headers can experience caching conflicts. For such applications that require full control over caching behavior, you must disable the persistence cookie.

Reviewer: ```suggestion
Applications which need control on `Cache-Control` headers (e.g. some which serve static content) can experience performance issues if they cannot properly rely on external caching. For such applications that require full control over caching behavior, you should consider disabling the persistence cookie.
```

Reviewer: Applications serving static content are the most usual example but not the only one. Performance issues are the biggest issues to highlight. The recommendation of disabling the persistence cookie should be lighter.

Author: :heavy_check_mark:

Reviewer: This is still pending

= Using cookies to keep route statefulness

[role="_abstract"]
To maintain stateful application traffic during pod restarts or scaling events, configure sticky sessions by using cookies. By using this method, you ensure that all incoming traffic reaches the same endpoint. This method helps prevent state loss even if the specific endpoint pod changes.

[IMPORTANT]
====

> The persistence cookie is enabled by default. Responses that contain the persistence cookie are automatically appended to the `Cache-Control: private` header to prevent the cookie from being cached. 

Reviewer: ```suggestion
The persistence cookie is enabled by default. Responses that contain the persistence cookie are automatically appended the `Cache-Control: private` header to prevent the cookie from being cached. 
```

Reviewer: The header is appended to the response, not the response to the header. Maybe I made some language error, please fix in that case, but it is the `Cache-Control: private` header what gets attached to the response.

Author: Thanks. Updated to:

> The persistence cookie is enabled by default. Responses that contain the persistence cookie automatically append the `Cache-Control: private` header to the reponse. This configuration prevents the cookie from being cached.

Reviewer: Still wrong. Responses are not the subject but the object. OpenShift Router appends the header to the response.

Author: :heavy_check_mark:

> The persistence cookie is enabled by default. Responses that contain the persistence cookie automatically append the `Cache-Control: private` header to the reponse. This configuration prevents the cookie from being cached. 

Reviewer: As said in the other comment (revamping here because GitHub does funny things when talking in resolved conversations): "Responses" should not be the subject. Basically, the OpenShift router appends the `Cache-Control: private`  header to the response, it is what needs to be clear.

Author: :heavy_check_mark:

> Red Hat appends the header. This happens because if the persistence cookie is cached, all clients receiving the cached response are routed to the same pod. This behavior can cause performance and reliability issues. 

Reviewer: ```suggestion
OpenShift Router appends the header because, if the persistence cookie is cached, all clients receiving the cached response are routed to the same pod. This behavior can cause performance and reliability issues. 
```

Reviewer: Key thing is that it is not Red Hat, as an organization, but OpenShift Router as an OCP component who adds the header.

Author: Thanks. Changed to:

> The {product-title} router appends the header to the response in situations where the persistence cookie is cached. Clients receiving the cached response are routed to the same pod. This cached behavior can cause performance and reliability issues.
